api/resource/resource.go

```diff
@@ -296,20 +295,37 @@ func (r *Resource) MustYaml() string {
 	return string(yml)
 }
 
+func (r *Resource) getOptions() *types.GenArgs {
+	annotations := r.GetAnnotations()
+	if genOptsAnno, ok := annotations[utils.BuildAnnotationsGenOptions]; ok {
+		var genOpts types.GeneratorArgs
+		yaml.Unmarshal([]byte(genOptsAnno), &genOpts)
+		return types.NewGenArgs(&genOpts)
+	}
+	return &types.GenArgs{}
+}
+
 // SetOptions updates the generator options for the resource.
 func (r *Resource) SetOptions(o *types.GenArgs) {
-	r.options = o
+	if o.IsNilOrEmpty() {
+		return
```

Comment:
Should this clear the annotation, e.g. is it possible it was present before and has been cleared during the lifetime of the Resource?

Reply:
I have added something to clear it when o == nil, but if I do so when o.IsNilOrEmpty, some generated resources don't get the hashes they need.

There's something very subtly different about o == nil vs o.args == nil, and the distinction is important so that most resources aren't hashed by default, but generated resources _are_ hashed by default. But I don't fully understand the code and it confuses me. In the case that o.args is empty, I tried clearing the annotation (which failed), and also setting the annotation to `{}`, which failed in a different way. 

The only time options must be cleared is during merging I think, when a generated resource is merging with an existing resource. `TestIssue3393` covers this case and the behavior is now corrected to what it was before.